micro-rdk/src/esp32/camera.rs

```diff
@@ -1,117 +1,278 @@
 #![allow(dead_code)]
-use std::time::Duration;
-
-use crate::common::camera::{Camera, CameraError};
-use crate::esp32::esp_idf_svc::sys::camera_config_t;
-use crate::esp32::esp_idf_svc::sys::camera_config_t__bindgen_ty_1;
-use crate::esp32::esp_idf_svc::sys::camera_config_t__bindgen_ty_2;
-use crate::esp32::esp_idf_svc::systime::EspSystemTime;
-use crate::proto::component::camera;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use crate::{
+    common::{
+        camera::{Camera, CameraError, CameraType},
+        config::ConfigType,
+        registry::{ComponentRegistry, Dependency},
+        status::{Status, StatusError},
+    },
+    esp32::esp_idf_svc::sys::{
+        camera::{
+            camera_config_t, camera_config_t__bindgen_ty_1, camera_config_t__bindgen_ty_2,
+            camera_fb_t, esp_camera_deinit, esp_camera_fb_get, esp_camera_fb_return,
+            esp_camera_init,
+        },
+        esp,
+    },
+    google::{self, api::HttpBody},
+    proto::component::camera,
+};
 use bytes::{Bytes, BytesMut};
-use log::*;
 use prost::Message;
+pub(crate) fn register_models(registry: &mut ComponentRegistry) {
+    if registry
+        .register_camera("esp32-camera", &Esp32Camera::from_config)
+        .is_err()
+    {
+        log::error!("esp32camera type is already registered");
+    }
+}
+
+static CAMERA_ALREADY_REGISTERED: Mutex<bool> = Mutex::new(false);
+
+enum PixelFormat {
+    /// 2BPP
+    RGB565 = 0,
+    /// 2BPP
+    YUV422 = 1,
+    /// 1.5BPP
+    YUV420 = 2,
+    /// 1BPP
+    GRAYSCALE = 3,
+    /// JPEG/Compressed
+    JPEG = 4,
+    /// 3BPP
+    RGB888 = 5,
+    RAW = 6,
+    /// 3BP2P
+    RGB444 = 7,
+    /// 3BP2P
+    RGB555 = 8,
+}
 
+/// Sizeof output image: QVGA|CIF|VGA|SVGA|XGA|SXGA|UXGA
+/// as u32.
+enum FrameSize {
+    /// 96x96, ~1.88KB JPEG
+    W96XH96 = 0,
+    /// 160x120, ~3.92 KB JPEG
+    QQVGA = 1,
+    /// 176x144, ~5.17 KB JPEG
+    QCIF = 2,
+    /// 240x176, ~8.62KB JPEG
+    HQVGA = 3,
+    /// 240x240, ~11.76KB JPEG
+    W240XH240 = 4,
+    /// 320x240, ~15.68 KB JPEG
+    QVGA = 5,
+    /// 400x296, ~24.18KB JPEG
+    CIF = 6,
+    /// 480x320, ~31.36KB JPEG
+    HVGA = 7,
+    /// 640x480, ~62.73KB JPEG
+    VGA = 8,
+}
+
+#[derive(DoCommand)]
 pub struct Esp32Camera {
+    active: bool,
```

Comment:
I don't actually see that `config` is ever used after `from_config` returns. Maybe this `struct` can just be empty?

Reply:
we will probably want to use it later for setting up the other camera methods, like getting the configured resolution and stuff. I haven't decided how I want to deal with it yet, will decide on it when implementing the other trait methods.